server/src/main/java/org/apache/lucene/search/uhighlight/BoundedBreakIteratorScanner.java

```diff
@@ -89,16 +89,33 @@ public int preceding(int offset) {
             innerStart = innerEnd;
             innerEnd = windowEnd;
         } else {
-            windowStart = innerStart = mainBreak.preceding(offset);
-            windowEnd = innerEnd = mainBreak.following(offset - 1);
-            // expand to next break until we reach maxLen
-            while (innerEnd - innerStart < maxLen) {
-                int newEnd = mainBreak.following(innerEnd);
-                if (newEnd == DONE || (newEnd - innerStart) > maxLen) {
-                    break;
+            innerStart = Math.max(mainBreak.preceding(offset), 0);
+
+            final int targetEndOffset = offset + Math.max(0, maxLen - (offset - innerStart));
+            final int textEndIndex = getText().getEndIndex();
+
+            if (targetEndOffset + 1 > textEndIndex) {
+                innerEnd = textEndIndex;
+            } else {
+                innerEnd = mainBreak.preceding(targetEndOffset + 1);
+            }
+
+            assert innerEnd != DONE && innerEnd >= innerStart;
+
+            // in case no break was found up to maxLen, find one afterwards.
+            if (innerStart == innerEnd) {
+                if (targetEndOffset >= textEndIndex) {
```

Comment:
You are right. The only way for `innerStart == innerEnd` to hold true for this case is when `text` is empty. In this case, the code would break without the condition. But given this is coded specifically for the unified highlighter, that doesn't seem like a possibility. Is that correct?